pkg/storage/storagewrappers/sharediterator/shared_iterator_datastore.go

```diff
@@ -196,76 +238,106 @@ func (sf *IteratorDatastore) ReadStartingWithUser(
 	}
 	span.SetAttributes(attribute.String("cache_key", cacheKey))
 
-	sf.internalStorage.mu.Lock()
-
-	keyItem, found := sf.internalStorage.iters[cacheKey]
+	// Check if the internal storage has reached its limit.
+	// If it has, we will bypass the shared iterator and use the inner reader directly.
+	// This is to prevent memory exhaustion and ensure that the storage does not grow indefinitely.
+	var count int
+	var full bool
+
+	// If the limit is zero, we will not use the shared iterator.
+	if sf.internalStorage.limit == 0 {
+		full = true
+	}
```

Comment:
```suggestion
	full = sf.internalStorage.limit == 0
```

Couldn't we just do this? 🤔

Reply:
yup, I can make that change across the datastore methods.